File: proto/python/cm2html.py
import commonmark
from pygments import highlight
from pygments.lexers import get_lexer_by_name
from pygments.formatters import HtmlFormatter
from pygments.util import ClassNotFound as LexerClassNotFound
import os, shutil, argparse

class CmdOption:
    def __init__(self):
        parser = argparse.ArgumentParser(description = "CommonMark to HTML converter.")
        parser.add_argument("in_dir", type = self.check_valid_dir)
        parser.add_argument("out_dir", nargs = "?", type = self.dir_chop)
        args = parser.parse_args()

        self.in_dir = args.in_dir
        self.out_dir = args.out_dir or os.getcwd()

    @staticmethod
    def dir_chop(val):
        val = val[:-1] if val.endswith('/') else val
        if not val:
            raise argparse.ArgumentTypeError("ルートディレクトリは指定できません (このエラーは回避可能ですが、通常そうすべきではありません…)")
        return val

    @staticmethod
    def check_valid_dir(val):
        if not os.path.isdir(val):
            raise argparse.ArgumentTypeError(f"ディレクトリ {val} が見つかりません")
        return CmdOption.dir_chop(val)

class CommonMark2HTML:
    def __init__(self):
        self.parser = commonmark.Parser()
        self.renderer = commonmark.HtmlRenderer()
        self.hl_fmtr = HtmlFormatter()
        self.doc = None

    def load(self, fh):
        md = fh.read()
        doc = self.parser.parse(md)
        for cur, entering in doc.walker():
            if cur.t == "code_block":
                language = cur.info
                try:
                    lexer = get_lexer_by_name(language)
                except LexerClassNotFound:
                    # 言語指定がない、または未対応の言語である場合は無視し、コードブロックのままにしておく
                    continue
                code_html = highlight(cur.literal, lexer, self.hl_fmtr)
                node = commonmark.node.Node("html_block", cur.sourcepos)
                node.literal = code_html
                cur.insert_before(node)
                cur.unlink()
        self.doc = doc

        return self

    def __str__(self):
        if not self.doc:
            return ""
        return self.renderer.render(self.doc)

    def get_style_defs(self):
        return self.hl_fmtr.get_style_defs('.highlight')

if __name__ == "__main__":
    cmdopt = CmdOption()
    os.makedirs(cmdopt.out_dir, exist_ok = True)
    conv = CommonMark2HTML()
    with open(os.path.join(cmdopt.out_dir, "style.css"), "w") as fout:
        fout.write(conv.get_style_defs())
    for cur, _, files in os.walk(cmdopt.in_dir):
        dir_offset = cur[len(cmdopt.in_dir) + 1:]
        dir_level = dir_offset.count('/') + 1 if dir_offset else 0
        os.makedirs(os.path.join(cmdopt.out_dir, dir_offset), exist_ok = True)
        for file in files:
            basename, ext = os.path.splitext(file)
            if ext.lower() == ".md":
                with open(os.path.join(cur, file), "r") as fin:
                    conv.load(fin)
                with open(os.path.join(cmdopt.out_dir, dir_offset, f"{basename}.html"), "w") as fout:
                    fout.write(f"""<!doctype html>
<html>
<head>
  <meta charset="utf-8">
  <title>{os.path.join(cur, file)}</title>
  <link rel="stylesheet" href="{''.join(['../'] * dir_level)}style.css">
</head>
<body>
{conv}
</body>
</html>
""")
            else:
                shutil.copy(os.path.join(cur, file), os.path.join(cmdopt.out_dir, dir_offset))

File: proto/js/cm2html.js
"use strict";

const fs = require("fs");
const commonmark = require("commonmark");
const hljs = require("highlight.js");

function CommonMark2HTML() {
  this.cmParser = new commonmark.Parser();
  this.cmWriter = new commonmark.HtmlRenderer();
}

CommonMark2HTML.prototype.load = function(file) {
  const md_src = fs.readFileSync(file);
  this.doc = this.cmParser.parse(md_src.toString());

  const walker = this.doc.walker();
  let item;
  while (item = walker.next()) {
    if (item.node.type !== "code_block") continue;
    const cur = item.node;
    let html;
    try {
      //TODO: フェンス情報の取扱については別途考慮すべき
      html = hljs.highlight(cur.info, cur.literal).value;
    }
    catch (ex) {
      // 未対応の言語の場合、コードブロックは手を付けない
      if (/^Error: Unknown language/.test(ex.toString())) continue;
      else throw ex;
    }
    const node = new commonmark.Node("html_block", cur.sourcepos);
    node.literal = '<pre class="highlight"><code>' + html + '</code></pre>';
    cur.insertBefore(node);
    cur.unlink();
  }
}

CommonMark2HTML.prototype.toString = function() {
  return this.cmWriter.render(this.doc);
}

function usage(err_msg) {
  if (!!err_msg) {
    console.error(err_msg);
  }
  console.info("usage: node cm2html.js in_dir out_dir");
  process.exit(!!err_msg ? 1 : 0);
}

if (process.argv.length != 4) {
  usage("引数が正しくありません。");
}
const [,,indir, outdir] = process.argv;
if (!fs.existsSync(indir) || !fs.statSync(indir).isDirectory()) {
  usage(`'$indir' はディレクトリではありません。`);
}
if (!fs.existsSync(outdir)) {
  try {
    fs.mkdirSync(outdir, { recursive: true })
  }
  catch (ex) {
    usage(`'$outdir' ディレクトリの作成に失敗しました。`);
  }
}
else {
  if (!fs.statSync(outdir).isDirectory()) {
    usage(`'$outdir' はディレクトリではありません。`);
  }
}

const parser = new commonmark.Parser();
const md = fs.readFileSync(process.argv[2]);
const doc = parser.parse(md.toString());

const walker = doc.walker();
let item;
while (item = walker.next()) {
  if (item.node.type !== "code_block") continue;
  const cur = item.node;
  let html;
  try {
    html = hljs.highlight(cur.info, cur.literal).value;
  }
  catch (ex) {
    if (/^Error: Unknown language/.test(ex.toString())) continue;
    else throw ex;
  }
  const node = new commonmark.Node("html_block", cur.sourcepos);
  node.literal = '<pre class="highlight"><code>' + html + '</code></pre>';
  cur.insertBefore(node);
  cur.unlink();
}

const renderer = new commonmark.HtmlRenderer();
const html_file = process.argv.length >= 4 ? process.argv[3] : process.argv[2] + ".html";
fs.writeFileSync(html_file, renderer.render(doc));
